NYCPlanning/db-pluto
# Store QAQC in Digital Ocean instead of Publishing Database

# Update storage of PLUTO QAQC Tables
This PR is to address issue #317 
Before you make any comments be sure to note that **this PR is not done yet**. Currently this PR is here to show the team  the work I've done thus far. There is a list of outstanding questions at the bottom, and once those are resolved then this can treated like a typical PR and eventually approved and merged. 

## How does PLUTO QAQC work?
The QAQC on this data product is different from our other data products in that new results are *added as rows to existing tables*. So the QAQC scripts need access to the results of the previous QAQC, which makes the process more complex.
Currently the results of QAQC on previous runs are all stored in the publishing database (the path to this DB is assigned to `EDM_DATA` in the `.env` files), and new results appended to these tables. 

## Why do we need an upgrade?
Currently the QAQC process writes directly to those 4 tables on the publishing database regardless of where it was run. This makes the info displayed on the [QAQC app](https://edm-data-engineering.nycplanningdigital.com/?page=Home) less reliable as QAQC on a testing run on someone's local machine could be displayed. Another issue is that one can't compares QAQC results run off different branches like one can for facDB. And making changes/additions to the QAQC scripts is extra hard as one has to be careful about messing with this important source of historical data. 

## New approach 
The new approach follows facDB's process of [writing QAQC tables to .csv's and putting them in an output folder that gets written to DO](https://github.com/NYCPlanning/db-facilities/blob/13447e36ce214bbc907c56b6f9d4740ef551e232/facdb/bash/export.sh#L11). This repo's implementation is more complex as the existing QAQC files must be pulled down first, and tables exported as .csv's for the app and .sql files for future builds. 

## Overview of changes to code

### Dataloading
Based this work off the [similar process in facDB](https://github.com/NYCPlanning/db-facilities/blob/13447e36ce214bbc907c56b6f9d4740ef551e232/facdb/bash/dataloading.sh#L6). One additional thing is to rename table with a SQL query written into the bash script

### QAQC Step
This was under the archive step, but I moved it to it's own bash script as I don't think QAQC and archive are very similar. 
I also refactored the code to loop through files and combinations of `CONDO` and `MAPPED` programmatically which made the code shorter and more readable. Previously the `CONDITION` was hardcoded and could be changed independently of the `CONDO` and `MAPPED` options which is quite fragile, but now it's set programmatically. 

### QAQC .sql files
Changes to these files is simple, point to tables in the local database and cast bbl's to float as types can get changed in the dataloading process

### Import qaqc function
This is based off the `import_public` function. We need this as the PLUTO QAQC process requires having QAQC on all versions/pairs in the same table

## Outstanding questions
- Is the `archive_pluto` the best table to use as "current"? There is also `export_pluto` and `pluto`
- Where should the qaqc .sql files be pulled from? Currently `05_qaqc.sql` points to the 317-QAQC-to_DO folder in `edm-publishing/db-pluto` but this is just for testing. I think that reading from the `latest` folder makes the most sense. This begs a larger question of whether all runs should upload files to the `latest` folder. We could implement something like facDB where files are uploaded to a folder named after the branch to avoid the common problem of test runs writing to the `latest` folder. This is preferable as it can be assumed that the output under `main` is correct.
- Where in the code should import_qaqc be called? I put in in `05_qaqc.sh` but it could live in the dataloading step as well. And where should it store the `.sql` files? I made a new sub-folder under `.library` but this is misleading as the data doesn't come from data library at all. It's also slightly confusing that the files are called `qaqc_expected.sql` when the code to *generate* that QAQC table has the same name. Curious if others find this confusing as well or if it's good enough

## How to test
You should be able to test without re-building. First run the new code in the dataloading step, either by commenting everything else out or running it from the command line. Once you have the previous pluto, you should be able to run the QAQC bash script on its own. Come to me of course with any questions on how to test this code locally.
